File: bindings/python/examples/ida2py_port.py
#!/usr/bin/env python3
"""
Headless Python adaptation of examples/tools/ida2py_port.cpp.

Usage:
    python examples/ida2py_port.py <binary-or-idb> [--list-user-symbols]
        [--show <name|ea>] [--cast <name|ea> <cdecl>] [--callsites <name|ea>]
        [--max-symbols <n>] [--appcall-smoke] [--quiet]
"""

from __future__ import annotations

import os
import sys
from dataclasses import dataclass, field

import idax


@dataclass(frozen=True)
class CastRequest:
    target: str
    declaration: str


@dataclass(frozen=True)
class Options:
    input: str
    quiet: bool = False
    list_user_symbols: bool = False
    show_targets: list[str] = field(default_factory=list)
    casts: list[CastRequest] = field(default_factory=list)
    callsites: list[str] = field(default_factory=list)
    appcall_smoke: bool = False
    max_symbols: int = 200


def is_idax_error(err: BaseException) -> bool:
    return isinstance(err, idax.IdaxError)


def error_message(err: BaseException) -> str:
    if isinstance(err, idax.IdaxError):
        context = f" ({err.context})" if err.context else ""
        return f"[{err.category}] {err}{context}"
    return str(err)


def hex_addr(address: int) -> str:
    return f"0x{address:x}"


def parse_address(text: str) -> int | None:
    trimmed = text.strip()
    if not trimmed:
        return None
    try:
        return int(trimmed, 0)
    except ValueError:
        return None


def resolve_symbol_or_address(token: str) -> int:
    direct = parse_address(token)
    if direct is not None:
        return direct
    return idax.name.resolve(token, idax.BAD_ADDRESS)


def parse_options(args: list[str]) -> Options:
    if not args:
        raise SystemExit("missing binary_file argument")

    input_path = args[0]
    quiet = False
    list_user_symbols = False
    appcall_smoke = False
    max_symbols = 200
    show_targets: list[str] = []
    callsites: list[str] = []
    casts: list[CastRequest] = []

    i = 1
    while i < len(args):
        arg = args[i]
        if arg in ("--help", "-h"):
            print(
                "Usage: python examples/ida2py_port.py <binary-or-idb> "
                "[--list-user-symbols] [--show <name|ea>] "
                "[--cast <name|ea> <cdecl>] [--callsites <name|ea>] "
                "[--max-symbols <n>] [--appcall-smoke] [--quiet]"
            )
            sys.exit(0)
        if arg in ("--quiet", "-q"):
            quiet = True
            i += 1
            continue
        if arg == "--list-user-symbols":
            list_user_symbols = True
            i += 1
            continue
        if arg == "--show":
            i += 1
            if i >= len(args):
                raise SystemExit("--show requires a value")
            show_targets.append(args[i])
            i += 1
            continue
        if arg == "--callsites":
            i += 1
            if i >= len(args):
                raise SystemExit("--callsites requires a value")
            callsites.append(args[i])
            i += 1
            continue
        if arg == "--cast":
            i += 1
            if i + 1 >= len(args):
                raise SystemExit("--cast requires <name|ea> <cdecl>")
            target = args[i]
            i += 1
            declaration = args[i]
            casts.append(CastRequest(target=target, declaration=declaration))
            i += 1
            continue
        if arg == "--max-symbols":
            i += 1
            if i >= len(args):
                raise SystemExit("--max-symbols requires a value")
            try:
                parsed = int(args[i])
            except ValueError:
                raise SystemExit("invalid --max-symbols value")
            if parsed <= 0:
                raise SystemExit("invalid --max-symbols value")
            max_symbols = parsed
            i += 1
            continue
        if arg == "--appcall-smoke":
            appcall_smoke = True
            i += 1
            continue
        raise SystemExit(f"unknown option: {arg}")

    if (
        not list_user_symbols
        and not show_targets
        and not callsites
        and not casts
        and not appcall_smoke
    ):
        list_user_symbols = True

    return Options(
        input=input_path,
        quiet=quiet,
        list_user_symbols=list_user_symbols,
        show_targets=show_targets,
        casts=casts,
        callsites=callsites,
        appcall_smoke=appcall_smoke,
        max_symbols=max_symbols,
    )


def list_user_symbols(max_symbols: int) -> None:
    entries = idax.name.all_user_defined(idax.BAD_ADDRESS, idax.BAD_ADDRESS)
    print("Address              Name                                Type")
    print("--------------------------------------------------------------------------")
    for entry in entries[:max_symbols]:
        type_name = "<none>"
        try:
            type_name = str(idax.types.retrieve(entry.address))
        except Exception:
            # No type at this symbol.
            pass
        print(f"{hex_addr(entry.address):<20} {entry.name:<34} {type_name}")


def inspect_symbol(token: str) -> None:
    ea = resolve_symbol_or_address(token)

    try:
        symbol_name = idax.name.get(ea)
    except Exception:
        symbol_name = "<unnamed>"

    refs_to = idax.xref.refs_to(ea)
    refs_from = idax.xref.refs_from(ea)
    preview = idax.data.read_bytes(ea, 16).hex(" ")

    print(f"\n== Show: {token} ==")
    print(f"address: {hex_addr(ea)}")
    print(f"name: {symbol_name}")

    try:
        demangled = idax.name.demangled(ea, idax.name.DemangleForm.short_)
        if len(demangled) > 0:
            print(f"demangled: {demangled}")
    except Exception:
        # Keep going.
        pass

    try:
        f = idax.function.at(ea)
        print(f"function: {f.name} [{hex_addr(f.start)} - {hex_addr(f.end)})")
    except Exception:
        # Not a function.
        pass

    try:
        ty = str(idax.types.retrieve(ea))
        print(f"type: {ty}")
    except Exception:
        # No type.
        pass

    print(f"bytes[16]: {preview}")
    print(f"xrefs_to: {len(refs_to)}")
    print(f"xrefs_from: {len(refs_from)}")


def apply_cast(request: CastRequest) -> None:
    ea = resolve_symbol_or_address(request.target)
    parsed = idax.types.from_declaration(request.declaration)
    parsed.apply(ea)
    roundtrip = str(idax.types.retrieve(ea))

    print(f"\n== Cast: {request.target} ==")
    print(f"address: {hex_addr(ea)}")
    print(f"applied: {request.declaration}")
    print(f"retrieved: {roundtrip}")


def show_callsites(target: str) -> None:
    callee = resolve_symbol_or_address(target)
    refs = [r for r in idax.xref.refs_to(callee) if r.is_code and idax.xref.is_call(r.type)]

    print(f"\n== Callsites: {target} ==")
    print(f"target: {hex_addr(callee)}")

    for ref in refs:
        try:
            caller = idax.function.at(ref.from_).name
        except Exception:
            caller = "<unknown>"
        try:
            line = idax.instruction.text(ref.from_)
        except Exception:
            line = "<decode failed>"

        print(f"  from {hex_addr(ref.from_)} ({caller}) -> {hex_addr(ref.to)} : {line}")

    print(f"callsites: {len(refs)}")


def main() -> None:
    options = parse_options(sys.argv[1:])

    try:
        idax.database.init(quiet=options.quiet)
        idax.database.open(options.input, True)
        idax.analysis.wait()

        if not options.quiet:
            print("== ida2py_port (Python adaptation) ==")
            print(f"input: {options.input}")
            print(f"processor: {idax.database.processor_name()}")
            print(f"address_bitness: {idax.database.address_bitness()}")

        if options.list_user_symbols:
            list_user_symbols(options.max_symbols)

        for target in options.show_targets:
            inspect_symbol(target)

        for cast in options.casts:
            apply_cast(cast)

        for target in options.callsites:
            show_callsites(target)

        if options.appcall_smoke:
            print("\n== Appcall smoke ==")
            print(
                "Appcall smoke is not exposed in this Python adaptation yet; "
                "use the C++ idax tool example for debugger-backed appcall validation."
            )
    except Exception as err:
        print(f"error: {error_message(err)}", file=sys.stderr)
        sys.exit(1)
    finally:
        try:
            idax.database.close(False)
        except Exception:
            # Ignore close errors during teardown.
            pass


if __name__ == "__main__":
    main()
    os._exit(0)  # avoid idalib shutdown crash

File: bindings/rust/idax/examples/ida2py_port.rs
mod common;

use common::{DatabaseSession, format_error, print_usage, resolve_symbol_or_address};
use idax::address::BAD_ADDRESS;
use idax::{Error, Result, data, database, function, instruction, name, types, xref};

#[derive(Debug, Clone)]
struct CastRequest {
    target: String,
    declaration: String,
}

#[derive(Debug, Clone)]
struct Options {
    input: String,
    quiet: bool,
    list_user_symbols: bool,
    show_targets: Vec<String>,
    casts: Vec<CastRequest>,
    callsites: Vec<String>,
    appcall_smoke: bool,
    max_symbols: usize,
}

impl Default for Options {
    fn default() -> Self {
        Self {
            input: String::new(),
            quiet: false,
            list_user_symbols: false,
            show_targets: Vec::new(),
            casts: Vec::new(),
            callsites: Vec::new(),
            appcall_smoke: false,
            max_symbols: 200,
        }
    }
}

fn parse_options(args: &[String]) -> Result<Options> {
    if args.len() < 2 {
        return Err(Error::validation("missing binary_file argument"));
    }

    let mut options = Options {
        input: args[1].clone(),
        ..Options::default()
    };

    let mut index = 2usize;
    while index < args.len() {
        match args[index].as_str() {
            "-h" | "--help" => {
                print_usage(
                    &args[0],
                    "<binary_file> [--list-user-symbols] [--show <name|ea>] [--cast <name|ea> <cdecl>] [--callsites <name|ea>] [--max-symbols <n>] [--appcall-smoke] [-q|--quiet]",
                );
                std::process::exit(0);
            }
            "-q" | "--quiet" => options.quiet = true,
            "--list-user-symbols" => options.list_user_symbols = true,
            "--show" => {
                index += 1;
                if index >= args.len() {
                    return Err(Error::validation("--show requires an argument"));
                }
                options.show_targets.push(args[index].clone());
            }
            "--cast" => {
                if index + 2 >= args.len() {
                    return Err(Error::validation(
                        "--cast requires <name|ea> <cdecl> arguments",
                    ));
                }
                options.casts.push(CastRequest {
                    target: args[index + 1].clone(),
                    declaration: args[index + 2].clone(),
                });
                index += 2;
            }
            "--callsites" => {
                index += 1;
                if index >= args.len() {
                    return Err(Error::validation("--callsites requires an argument"));
                }
                options.callsites.push(args[index].clone());
            }
            "--max-symbols" => {
                index += 1;
                if index >= args.len() {
                    return Err(Error::validation("--max-symbols requires a numeric value"));
                }
                options.max_symbols = args[index]
                    .parse::<usize>()
                    .map_err(|_| Error::validation("invalid --max-symbols value"))?;
            }
            "--appcall-smoke" => options.appcall_smoke = true,
            unknown => return Err(Error::validation(format!("unknown option: {unknown}"))),
        }
        index += 1;
    }

    if !options.list_user_symbols
        && options.show_targets.is_empty()
        && options.casts.is_empty()
        && options.callsites.is_empty()
        && !options.appcall_smoke
    {
        options.list_user_symbols = true;
    }

    Ok(options)
}

fn list_user_symbols(max_symbols: usize, output: &mut String) -> Result<()> {
    let entries = name::all_user_defined(BAD_ADDRESS, BAD_ADDRESS)?;
    output.push_str("\n== User-defined Symbols ==\n");
    output.push_str("Address              Name                                Type\n");
    output.push_str("--------------------------------------------------------------------------\n");

    for entry in entries.iter().take(max_symbols) {
        let type_name = types::retrieve(entry.address)
            .ok()
            .and_then(|ty| ty.to_string().ok())
            .unwrap_or_else(|| "<none>".to_string());
        output.push_str(&format!(
            "{:<20} {:<34} {}\n",
            format!("0x{:x}", entry.address),
            entry.name,
            type_name
        ));
    }
    Ok(())
}

fn inspect_symbol(target: &str, output: &mut String) -> Result<()> {
    let address = resolve_symbol_or_address(target)?;
    let symbol_name = name::get(address).unwrap_or_else(|_| "<unnamed>".to_string());
    let demangled = name::demangled(address, name::DemangleForm::Short).unwrap_or_default();
    let refs_to = xref::refs_to(address).unwrap_or_default();
    let refs_from = xref::refs_from(address).unwrap_or_default();
    let bytes = data::read_bytes(address, 16).unwrap_or_default();

    output.push_str(&format!("\n== Show: {target} ==\n"));
    output.push_str(&format!("address: 0x{:x}\n", address));
    output.push_str(&format!("name: {symbol_name}\n"));
    if !demangled.is_empty() {
        output.push_str(&format!("demangled: {demangled}\n"));
    }

    if let Ok(function_info) = function::at(address) {
        output.push_str(&format!(
            "function: {} [{} - {})\n",
            function_info.name(),
            format!("0x{:x}", function_info.start()),
            format!("0x{:x}", function_info.end())
        ));
    }

    if let Ok(ty) = types::retrieve(address)
        && let Ok(ty_text) = ty.to_string()
    {
        output.push_str(&format!("type: {ty_text}\n"));
    }

    let preview = bytes
        .iter()
        .map(|byte| format!("{byte:02x}"))
        .collect::<Vec<String>>()
        .join(" ");
    output.push_str(&format!("bytes[16]: {preview}\n"));
    output.push_str(&format!("xrefs_to: {}\n", refs_to.len()));
    output.push_str(&format!("xrefs_from: {}\n", refs_from.len()));
    Ok(())
}

fn apply_cast(request: &CastRequest, output: &mut String) -> Result<()> {
    let address = resolve_symbol_or_address(&request.target)?;
    let parsed = types::TypeInfo::from_declaration(&request.declaration)?;
    parsed.apply(address)?;
    let roundtrip = types::retrieve(address)?.to_string()?;

    output.push_str(&format!("\n== Cast: {} ==\n", request.target));
    output.push_str(&format!("address: 0x{:x}\n", address));
    output.push_str(&format!("applied: {}\n", request.declaration));
    output.push_str(&format!("retrieved: {roundtrip}\n"));
    Ok(())
}

fn show_callsites(target: &str, output: &mut String) -> Result<()> {
    let target_address = resolve_symbol_or_address(target)?;
    let references = xref::refs_to(target_address)?;

    output.push_str(&format!("\n== Callsites: {target} ==\n"));
    output.push_str(&format!("target: 0x{:x}\n", target_address));

    let mut call_count = 0usize;
    for reference in references {
        if !reference.is_code || !xref::is_call(reference.ref_type) {
            continue;
        }
        call_count += 1;
        let caller = function::at(reference.from)
            .map(|f| f.name().to_string())
            .unwrap_or_else(|_| "<unknown>".to_string());
        let line =
            instruction::text(reference.from).unwrap_or_else(|_| "<decode failed>".to_string());
        output.push_str(&format!(
            "  from 0x{:x} ({caller}) -> 0x{:x} : {line}\n",
            reference.from, reference.to
        ));
    }
    output.push_str(&format!("callsites: {call_count}\n"));
    Ok(())
}

fn run() -> Result<()> {
    let args: Vec<String> = std::env::args().collect();
    let options = parse_options(&args)?;

    let _session = DatabaseSession::open(&options.input, true)?;

    let mut output = String::new();
    if !options.quiet {
        output.push_str("== ida2py_port (Rust adaptation) ==\n");
        output.push_str(&format!("input: {}\n", options.input));
        output.push_str(&format!(
            "processor: {}\n",
            database::processor_name().unwrap_or_else(|_| "<unknown>".to_string())
        ));
        output.push_str(&format!(
            "address_bitness: {}\n",
            database::address_bitness().unwrap_or_default()
        ));
    }

    if options.list_user_symbols {
        list_user_symbols(options.max_symbols, &mut output)?;
    }

    for target in &options.show_targets {
        inspect_symbol(target, &mut output)?;
    }

    for cast in &options.casts {
        apply_cast(cast, &mut output)?;
    }

    for target in &options.callsites {
        show_callsites(target, &mut output)?;
    }

    if options.appcall_smoke {
        output.push_str(
            "\n== Appcall smoke ==\nAppcall smoke is not exposed in this safe Rust example surface yet; use C++ idax tool examples for debugger-backed appcall validation.\n",
        );
    }

    print!("{output}");
    Ok(())
}

fn main() {
    if let Err(error) = run() {
        eprintln!("error: {}", format_error(&error));
        std::process::exit(1);
    }
}
